arch/arm/core/aarch32/prep_c.c

```diff
@@ -37,7 +37,7 @@
 
 #include <string.h>
 
-#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY)
```

Comment:
This is not good. You cannot skip VTOR initialization on a device that has VTOR. Even if you build a Bootloader, you can't guarantee that this is the first image in boot sequence, so VTOR needs to be determined.

Reply:
Also, if this is required, particularly, if the relay handler is not at the beginning of flash. There are SoCs that require special sections before the ROM vector table, so VTOR requires to be set.

Reply:
so we want that only bootloader touch VTOR.

Reply:
#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR

#ifdef CONFIG_XIP
#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
#else
#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
#endif

#else

#if defined(CONFIG_SW_VECTOR_RELAY)
Z_GENERIC_SECTION(.vt_pointer_section) void *_vector_table_pointer;
#endif

#define VECTOR_ADDRESS 0
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */



``` C
#if CONFIG_CPU_CORTEX_M_HAS_VTOR
static inline void relocate_vector_table(void)
#else
void __weak relocate_vector_table(void)
#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
{
#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_BOOTLOADER_MCUBOOT)
	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
	__DSB();
	__ISB();
#endif
#if !defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) || defined(CONFIG_SW_VECTOR_RELAY)
#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
#endif
}
```
	__DSB();
	__ISB();

#if defined(CONFIG_XIP) && (CONFIG_FLASH_BASE_ADDRESS != 0) || \
    !defined(CONFIG_XIP) && (CONFIG_SRAM_BASE_ADDRESS != 0)
	size_t vector_size = (size_t)_vector_end - (size_t)_vector_start;
	(void)memcpy(VECTOR_ADDRESS, _vector_start, vector_size);
#elif defined(CONFIG_SW_VECTOR_RELAY)
	_vector_table_pointer = _vector_start;
#endif
}

Reply:
I have applied the suggestion with one major change. The relay table address is under `__vector_relay_table` variable, therefore `VECTOR_ADDRESS` was set accordingly.